### PR TITLE
General maintenance for Flat Packer

### LINKED ISSUE #84150: Flatpacker lets you stack machines ontop of eachother
Reporting client version: 515.1631

## Issue Summary
Flatpacker packs do not check if a tile is already occupied by something dense when you unpack them, allowing machines to be stacked onto the same tile, this is particularly powerful with modular shield generators because you can make an infinitely large shield generator in a 2 tile space

## Round ID:
[230812](https://scrubby.melonmesa.com/round/230812)

## Testmerges:
- [[TM ONLY] Hijacked Revision Date for v3 TGS Api](https://www.github.com/tgstation/tgstation/pull/83227)

## Reproduction:
Pack machines with flat packer

Stack packs on the same tile

multi tool them all

image

### LINKED ISSUE #84188: Poor flatpacker feedback
Reporting client version: 515.1636

## Issue Summary

Very minor consistency issue. The flatpacked box doesn't say anywhere you can't use the multitool on the flatpacked box if it's in your hand. Should update the balloon alert & item description to say 'can't use in hand!' as most items like that are usually used in hand.

## Round ID:
[230879](https://scrubby.melonmesa.com/round/230879)

## Testmerges:
- [[TM ONLY] Hijacked Revision Date for v3 TGS Api](https://www.github.com/tgstation/tgstation/pull/83227)

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
**1. Qol**
 - Adds examines, screen tips for screwdriver & crowbar acts for flat packer
 - Adds examines, balloon alerts for when a flat packed box can't be deployed (Closes #84188)
 - Adds screen tip for flatpack cart
 - Flatpack shows the progress bar animation on its icon during printing to let us know that it's doing some work
 - Eject button on flat packer UI is disabled when no circuit board is inserted
 - Ejecting the board will place it in your hand, use ctrl click shortcut to eject board

 **2. Code Improvements**
  - Improved `analyze_cost()`. It only creates a temporary `null_comp` if the materials are not available on the techweb saving us from some extra gc work whenever this proc is called
  - Merged `start_build()` with `ui_act()` to save us an extra proc call
  - Improved `ui_data()` to compute no data when no circuit board is inserted
  - Just re arranged code everywhere to make it look clean in my opinion atleast

**3. Fixes**
- Inserting different materials will display the correct colour insertion animation(as was implemented in #81692)
- Opening the panel of a flat packer with a screwdriver now has an open icon state
- You can deconstruct a flat packer with a crowbar after opening its maintenance panel
- Fixes #84150. Flat packed boxes now have density & won't deploy on a turf that is blocked by dense objects.
- Correct cost of the design scaled with `creation_coefficiency` is displayed in the UI
- No hologram & abstract item interactions with the flat packer & the flatpack box. You can also hit them with objects like a flatpack box, screwdriver, crowbar etc when in combat mode

## Changelog
:cl:
qol: adds examines, screentips & balloon alerts for flatpacker, flatpack box & cart
qol: adds correct material colour insertion animation for flat packer, continuous progress bar animation during printing
qol: flat packer has a separate icon when you open its maintenance panel, eject button is disabled when no board is inserted
qol: ejecting the board will place it in your hand, use ctrl click shortcut to eject board from flatpacker
fix: you cannot deploy a flat packed box on a turf that has dense objects(like other flat packed boxes & such)
fix: you cannot insert hologram/abstract items into a flat packer or flatpack cart, also can hit them with any objects when in combat mode
fix: you can deconstruct a flat packer with a crowbar after opening its maintenance panel
fix: correct cost of the design scaled with `creation_coefficiency` is displayed in the UI
code: overall improved code quality of all things flat pack related
/:cl:

### CODE REVIEW

File: code/game/machinery/flatpacker.dm [hunk 1]
```diff
@@ -136,189 +178,247 @@
...
 	name += " ([initial(build.name)])"
 
-
 /obj/item/flatpack/Destroy()
 	QDEL_NULL(board)
 	. = ..()
 
+/obj/item/flatpack/examine(mob/user)
+	. = ..()
+	if(in_range(user, src) || isobserver(user))
```
Comment (jlsnow301) [inline code/game/machinery/flatpacker.dm:350]: early pls

File: code/game/machinery/flatpacker.dm [hunk 2]
```diff
@@ -34,10 +36,57 @@
...
+	if(held_item.tool_behaviour == TOOL_SCREWDRIVER)
+		context[SCREENTIP_CONTEXT_LMB] = "[panel_open ? "Close" : "Open"] panel"
+		return CONTEXTUAL_SCREENTIP_SET
+	else if(held_item.tool_behaviour == TOOL_CROWBAR && panel_open)
+		context[SCREENTIP_CONTEXT_LMB] = "Deconstruct"
+		return CONTEXTUAL_SCREENTIP_SET
+
+/obj/machinery/flatpacker/examine(mob/user)
+	. += ..()
+	if(in_range(user, src) || isobserver(user))
```
Comment (jlsnow301) [inline code/game/machinery/flatpacker.dm:63]: early

**Review by jlsnow301 — COMMENTED** [review]
removing double spacing between procs 😢 wait till ya see how i'm structuring my dm now

**Comment by SyncIt21** [comment]
This PR is missing all it's labels. @jlsnow301 hope this I'm not bothering you but could  you add them back, please :) ?

**Pull request merged**